*  DSC 530
*  Week 3 & 4 Assignment
*  Peter Lozano

# Hands-on data analysis with Pandas

# Chapter 1, Exercise 4

Run the code in the first cell of the exercises.ipynb notebook. It will give you a list of 100 values to work with for the rest of the exercises in this chapter. Be sure to treat these values as a sample of the population.


In [2]:
import random

random.seed(0)
salaries = [round(random.random()*1000000, -3) for _ in range(100)]
# Currently, this code generated a list of random values as salaries for this exercise.

## Chapter 1, Exercise 5

Using the data from exercise 4, calculate the following statistics without importing anything from the [statistics module](https://docs.python.org/3/library/statistics.html) in the Python standard library, and then confirm your results match up to those that are obtained when using the statistics module (where possible):


Tip from instructor: You can use [Pandas](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html) or [Numpy](https://numpy.org/doc/stable/reference/routines.statistics.html) for the initial calculation and then check against other Python library/package such as [statistics module](https://docs.python.org/3/library/statistics.html), or [SciPy Statistics](https://docs.scipy.org/doc/scipy/reference/stats.html). Part a of Exercise 5 is completed to illustrate the process for you, you can decide how to go about the other problems.

Exercise 5(a) Mean - with exmaples

In [3]:
# Run this code to make alias to shorted the typed name
# You can do this as the 1st thing in any notebook or later, but you only need to do this once
import numpy as np
import pandas as pd

In [4]:
# Using numpy.mean
np.mean(salaries)

np.float64(585690.0)

In [5]:
# Using statistics module
import statistics # no need to re-run this after import
# You have to import this standard library before using its mean function
statistics.mean(salaries)

585690.0

In [6]:
# Using Pandas
# If you choose to use Pandas functions, you have to convert the list to a Series, otherwise it will gives an error
salaries_series = pd.Series(salaries)
# For Pandas function, when you placed a dot after a variable or other functions, it is referred as a "method chaining" appraoch.
salaries_series.mean()

np.float64(585690.0)

In [7]:
# Using SciPy Statistics
from scipy import stats
stats.tmean(salaries)

np.float64(585690.0)

Exercise 5(b) Median

In [8]:
# Your 1st approach
# Median calculation
sorted_salaries = sorted(salaries)
n = len(sorted_salaries)
# Using if else to check if n is odd or even
if n % 2 == 1:
    median = sorted_salaries[n // 2]
else:
    median = (sorted_salaries[n // 2 - 1] + sorted_salaries[n // 2]) / 2

median

589000.0

In [9]:
# Using statistics module

median = statistics.median(salaries)

median

589000.0

Exercise 5(c) Mode

Hint from textbook: Check out the [Counter class](https://docs.python.org/3/library/collections.html#collections.Counter) in the collections module of the standard library

In [10]:
# Your 1st approach
# Mod calculation using counter class
from collections import Counter
counts = Counter(salaries)
mode = counts.most_common(1)[0][0]
mode

477000.0

In [11]:
# Using statistics module
mode = statistics.mode(salaries)

mode

477000.0

Exercise 5(d) Sample Variance

In [12]:
# Your 1st approach
# Count observations
n = len(salaries)
# Divide sum of salaries by n
mean = sum(salaries) / n
# Subtract mean from each salary, square the result, and sum all squared differences
squared_diffs = [(x - mean) ** 2 for x in salaries]
variance = sum(squared_diffs) / n

variance

69957413900.0

In [13]:
# Using statistics module
variance = statistics.variance(salaries)
variance

# Variance is different from manual calculation because statistics.variance 
# uses n-1 (sample variance) in the denominator by default.

70664054444.44444

In [14]:
# Inserting 2nd approach for validation purposes
# Count observations
n = len(salaries)
# Divide sum of salaries by n
mean = sum(salaries) / (n - 1)
# Subtract mean from each salary, square the result, and sum all squared differences
squared_diffs = [(x - mean) ** 2 for x in salaries]
# Divide sum of salaries by n - 1 to reflect statistics.variance behavior
variance = sum(squared_diffs) / (n - 1)

variance

70699407750.60059

Exercise 5(e) Sample Standard Deviation

In [15]:
# Your 1st approach
# Sample Standard Deviation calculation
n = len(salaries)
# Divide sum of salaries by n
mean = sum(salaries) / n
# Subtract mean from each salary, square the result, and sum all squared differences
squared_diffs = [(x - mean) ** 2 for x in salaries]
# Divide sum of squared differences by n - 1 to get sample variance
variance = sum(squared_diffs) / (n - 1)
# Take square root of variance to get standard deviation
std_dev = variance ** 0.5

std_dev

265827.11382484

In [16]:
# Using statistics module
std_dev = statistics.stdev(salaries)
    
std_dev

265827.11382484

## Chapter 1, Exercise 6

Using the data from exercise 4, calculate the following statistics using the functions in the statistics module where appropriate:

Exercise 6(a) Range

In [17]:
# Hint:there is no function called "Range". However, what does it means to see a range of values?
data_range = max(salaries) - min(salaries)
data_range

995000.0

Exercise 6(b) Coefficient of Variation

In [18]:
# Hint: Coefficient of Variation (CV) is defined as the ratio of standard deviation to the mean in Hands-on Data Analysis with Pandas.
# Coefficient of Variation is a ratio of the standard deviation to the mean
# It measures the extent of variability in relation to the mean of the population.
# The higher the CV, the greater the level of dispersion around the mean.
cv = std_dev / statistics.mean(salaries)
cv

0.45386998894439035

Exercise 6(c) Interquartile Range

In [19]:
# Hint: Interquartile range (IQR) is presented as the difference of the 75% and 25% quantile in Chapter 1 of Practical Statistics for Data Scientisits Textbook
# OR you can find the values in each quartile and then substract Quantile 3 with Quantile 1 as in Chapter 1 of Hands-on Data Analysis with Pandas
# Extracting the 3rd quartile using numpy.percentile
q3 = np.percentile(salaries, 75)
# Extracting the 1st quartile using numpy.percentile
q1 = np.percentile(salaries, 25)
iqr = q3 - q1
iqr

np.float64(413250.0)

Exercise 6(d) Quartile coefficient of dispersion

In [20]:
# Once you determined the 1st and 3rd quantile values from 6(c), you can use it for this calculation
# Refer to the Quartile coefficient of dispersion formula in Chapter 1 of Hands-on Data Analysis with Pandas
qcd = (q3 - q1) / (q3 + q1)
qcd

np.float64(0.338660110633067)

## Chapter 1, Exercise 7

Scale the data created in exercise 4 using the following strategies:

Exercise 7(a) Min-max Scaling (Normalizing)

In [21]:
# Refer to the Scaling data section of your reading in Chapter 1 of Hands-on Data Analysis with Pandas
# You can calculate this value by following the formula in the textbook or see the example in the solution
# You can also test out the scikit-learn function "MinMaxScaler", which requires transforming the data in 2D array before applying the scaler. This is a common approach in machine learning.

# Solution to display min-max scaled salaries (Normalization)
min_salary = min(salaries)
max_salary = max(salaries)
scaled_salaries_manual = [(x - min_salary) / (max_salary - min_salary) for x in salaries]
scaled_salaries_manual

# Testing with sklearn MinMaxScaler
# from sklearn.preprocessing import MinMaxScaler
# scaler = MinMaxScaler()
# salaries_reshaped = np.array(salaries).reshape(-1, 1)
# scaled_salaries = scaler.fit_transform(salaries_reshaped)

[0.8472361809045226,
 0.7608040201005025,
 0.4221105527638191,
 0.2592964824120603,
 0.5125628140703518,
 0.40603015075376886,
 0.7869346733668342,
 0.3035175879396985,
 0.47839195979899496,
 0.5849246231155779,
 0.9115577889447236,
 0.5065326633165829,
 0.28241206030150756,
 0.7587939698492462,
 0.6201005025125628,
 0.25125628140703515,
 0.91356783919598,
 0.9869346733668342,
 0.8130653266331658,
 0.9055276381909547,
 0.31055276381909547,
 0.7326633165829146,
 0.9025125628140703,
 0.6864321608040201,
 0.4733668341708543,
 0.10050251256281408,
 0.43517587939698493,
 0.6130653266331658,
 0.9165829145728643,
 0.9708542713567839,
 0.47839195979899496,
 0.8683417085427135,
 0.26030150753768844,
 0.8080402010050252,
 0.5507537688442211,
 0.01306532663316583,
 0.7226130653266332,
 0.4,
 0.828140703517588,
 0.6703517587939698,
 0.0,
 0.49547738693467336,
 0.871356783919598,
 0.2442211055276382,
 0.3256281407035176,
 0.8733668341708543,
 0.19095477386934673,
 0.5698492462311557,
 0.23919597989

Exercise 7(b) Standardizing

In [22]:
# Refer to the Scaling data section of your reading in Chapter 1 of Hands-on Data Analysis with Pandas
# You can calculate this value by following the formula in the textbook or see the example in the solution
# You can also test out the z-score method in Scipy Stats (2D array is needed)

# Solution to display z-score standardized salaries (Standardization)
mean_salary = statistics.mean(salaries)
std_dev_salary = statistics.stdev(salaries)
standardized_salaries_manual = [(x - mean_salary) / std_dev_salary for x in salaries]
# Rounding to 8 decimal places for better readability
standardized_salaries_manual = [round(x, 8) for x in standardized_salaries_manual]

# Testing with scipy stats zscore
from scipy import stats
salaries_reshaped = np.array(salaries).reshape(-1, 1)
standardized_salaries = stats.zscore(salaries_reshaped)

# Note: The result from scipy.stats.zscore is slightly different from manual calculation due to precision of floating point arithmetic.

# Printing both standardized salaries for comparison
print("Standardized Salaries (Manual Calculation):")
print(standardized_salaries_manual)
print("\nStandardized Salaries (SciPy Stats zscore):")
print(standardized_salaries)

Standardized Salaries (Manual Calculation):
[0.97172179, 0.64820325, -0.61953801, -1.22895665, -0.28097209, -0.6797275, 0.74601118, -1.06343554, -0.40887477, -0.01011936, 1.21247978, -0.30354315, -1.14243425, 0.64067957, 0.12154516, -1.2590514, 1.22000347, 1.49461804, 0.84381912, 1.18990872, -1.03710263, 0.54287164, 1.17862319, 0.36982683, -0.42768399, -1.82332793, -0.57063404, 0.09521226, 1.231289, 1.43442854, -0.40887477, 1.05072051, -1.22519481, 0.8250099, -0.13802204, -2.15060831, 0.5052532, -0.70229856, 0.90024677, 0.30963734, -2.19951228, -0.34492343, 1.06200604, -1.28538431, -0.98067498, 1.06952973, -1.48476201, -0.06654701, -1.30419352, 1.43819039, 0.81748621, -0.51796823, -1.90232664, -0.9994842, -0.29225762, 1.30652587, -1.79323318, -0.13049835, 0.45634923, -0.14554572, 0.85886649, -0.17187863, 1.42314301, 0.06511751, 0.00868986, -0.52925376, 0.03878461, -0.75496437, -0.03645226, -1.1123395, -1.4922857, -1.49980938, 0.10273595, 0.26825706, -0.40887477, -1.8647082, 0.64820325,

## Chapter 1, Exercise 8

Using the scaled data from exercise 7, calculate the following:

Note. There is some preparation work to create the normalized (scaled) vs. standardized dataset before calculating covariance.

In [23]:
# If you did not setup a scaled data or standardized data with z-score approach, you can use the following codes.
# Scaled data aka normalized data
from sklearn.preprocessing import MinMaxScaler
# Make an array of salaries
salaries_array = np.array(salaries)
# Reshape the data to 2D (one feature)
salaries_2d = salaries_array.reshape(-1, 1)
# Scaled the data
scaler = MinMaxScaler()
scaler.fit(salaries_2d)
# Optionally, transform the data. scaled_salaries in the scaled data for use in next step
scaled_salaries = scaler.transform(salaries_2d)
# Preview the first 10 values. Noticed the double [] that shows the 2D shape
scaled_salaries[:10]
# This will print the shape if you would like to check
# scaled_salaries.shape

array([[0.84723618],
       [0.76080402],
       [0.42211055],
       [0.25929648],
       [0.51256281],
       [0.40603015],
       [0.78693467],
       [0.30351759],
       [0.47839196],
       [0.58492462]])

In [24]:
# Standardized version of the salary data (z-score)
from scipy import stats
# Assign a new variable `standardized_salary` to store the standarized data
standardized_salary = stats.zscore(salaries_array)
# Preview the first 10 values. This is a 1 dimension object
standardized_salary[:10]
# This will print the shape if you would like to check
# standardized_salary.shape

array([ 0.97661715,  0.65146878, -0.62265912, -1.23514791, -0.28238758,
       -0.68315184,  0.74976945, -1.06879293, -0.41093461, -0.01017034])

Exercise 8(a) The covariance between the standardized and normalized data

In [25]:
# Covariance can be calculated with numpy
# When the (normalized) scaled_salaries varaible was created, it was created as a 2D array.
# We will have to 'flatten' it to match the shape of the standardized_salary before calculation.
scaled_salaries = np.array(scaled_salaries).flatten()

# Now calculate the covariance
cov_calculated = np.cov(scaled_salaries, standardized_salary)
print(cov_calculated)

[[0.07137603 0.26850885]
 [0.26850885 1.01010101]]


Note: How to interpret this output? The covariance function produce a 2X2 covariance matrix, the off-diagonal elements present the covariance of the 2 sets of values. Therefore 0.2685 is the output for intrepretation. Since the covariance estimate is positive, it indicates a positive relationship (moving in the same direction), and we don't know how "strong" this relationship and therefore correlation will help us understand this better.

In [26]:
# Example 2 for covariance calculation
# Covariance can be calculated with statistics module
statistics.covariance(scaled_salaries, standardized_salary)

np.float64(0.2685088459449036)

Exercise 8(b) The Pearson correlation coefficient between the standardized and normalized data (this is actually 1, but due to rounding along the way, the result will be slightly less)

In [27]:
# Correlation can be calculated with statistics module
statistics.correlation(scaled_salaries, standardized_salary)
# The result is 1 because we are checking relationship of the same data in different format.

np.float64(1.0000000000000002)

Note. Getting a correlation of 1 is often very rare, but in this case, we are basically using the same raw data and simply normalized it (force values to 0,1 range) or standardized it (force the data to centered around zero and then deviate from zero) to put it on a common scale. It will not change the strength or direction of the relationship between 2 variables.

# Chapter 2

In [28]:
# The default code to run if you setup your book_env correctly to import the dataset
# Just replaced an additional ../ to the path
df = pd.read_csv('../ch_02/data/parsed.csv')

# If you encountered issues due to directory setup, you can provide the full file path also
# df = pd.read_csv('cdrive../../ch_02/data/parsed.csv')

df.head()

,alert,cdi,code,detail,dmin,felt,gap,ids,mag,magType,...,status,time,title,tsunami,type,types,tz,updated,url,parsed_place
0,NaN,NaN,37389218,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.008693,NaN,85.0,",ci37389218,",1.35,ml,...,automatic,1539475168010,"M 1.4 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475395144,https://earthquake.usgs.gov/earthquakes/eventp...,California
1,NaN,NaN,37389202,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.020030,NaN,79.0,",ci37389202,",1.29,ml,...,automatic,1539475129610,"M 1.3 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475253925,https://earthquake.usgs.gov/earthquakes/eventp...,California
2,NaN,4.4,37389194,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.021370,28.0,21.0,",ci37389194,",3.42,ml,...,automatic,1539475062610,"M 3.4 - 8km NE of Aguanga, CA",0,earthquake,",dyfi,focal-mechanism,geoserve,nearby-cities,o...",-480.0,1539536756176,https://earthquake.usgs.gov/earthquakes/eventp...,California
3,NaN,NaN,37389186,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.026180,NaN,39.0,",ci37389186,",0.44,ml,...,automatic,1539474978070,"M 0.4 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475196167,https://earthquake.usgs.gov/earthquakes/eventp...,California
4,NaN,NaN,73096941,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.077990,NaN,192.0,",nc73096941,",2.16,md,...,automatic,1539474716050,"M 2.2 - 10km NW of Avenal, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,scit...",-480.0,1539477547926,https://earthquake.usgs.gov/earthquakes/eventp...,California


## Chapter 2, Exercise 1

Find the 95th percentile of earthquake magnitude in Japan using the mb magnitude type.

In [29]:
# Finding the 95th percentile of the earthquake magnitudes in Japan
# --- IGNORE ---
japan_earthquakes = df[df['place'].str.contains('Japan', na=False)]
japan_magnitudes = japan_earthquakes['mag'].dropna()
np.percentile(japan_magnitudes, 95)

np.float64(5.119999999999999)

## Chapter 2, Exercise 2

Find the percentage of earthquakes in Indonesia that were coupled with tsunamis.

In [30]:
# Finding the percentage of earthquakes in Indonesia that were coupled with a tsunami
indonesia_earthquakes = df[df['place'].str.contains('Indonesia', na=False)]
# Filtering data where 'tsunami' column is 1 indicating a tsunami occurred
indonesia_tsunami = indonesia_earthquakes[indonesia_earthquakes['tsunami'] == 1]
# Calculating the percentage of earthquakes that were accompanied by a tsunami
percentage_tsunami = (len(indonesia_tsunami) / len(indonesia_earthquakes)) * 100 if len(indonesia_earthquakes) > 0 else 0
percentage_tsunami



23.12925170068027

## Chapter 2, Exercise 3

Calculate summary statistics for earthquakes in Nevada.

In [31]:
# Calculate summary statistics for earthquakes in Nevada
# --- IGNORE ---
nevada_earthquakes = df[df['place'].str.contains('Nevada', na=False)]
nevada_magnitudes = nevada_earthquakes['mag'].dropna()
nevada_summary = {
    'mean': nevada_magnitudes.mean(),
    'median': nevada_magnitudes.median(),
    # Mode might not be unique, so we handle it with a try-except block
    'mode': nevada_magnitudes.mode().iloc[0] if not nevada_magnitudes.mode().empty else None,
    # Calculating Mode might not be unique, so we handle it with a try-except block
    'std_dev': nevada_magnitudes.std(),
    'variance': nevada_magnitudes.var(),
    'min': nevada_magnitudes.min(),
    'max': nevada_magnitudes.max()
}
nevada_summary

{'mean': np.float64(0.4917282127031019),
 'median': np.float64(0.4),
 'mode': np.float64(0.1),
 'std_dev': np.float64(0.6895604444106154),
 'variance': np.float64(0.47549360649576533),
 'min': np.float64(-0.5),
 'max': np.float64(2.9)}

## Chapter 2, Exercise 4

Add a column indicating whether the earthquake happened in a country or US state that is on the Ring of Fire. Use Alaska, Antarctica (look for Antarctic), Bolivia, California, Canada, Chile, Costa Rica, Ecuador, Fiji, Guatemala, Indonesia, Japan, Kermadec Islands, Mexico (be careful not to select New Mexico), New Zealand, Peru, Philippines, Russia, Taiwan, Tonga, and Washington.

In [32]:
# Adding a column to indicate whether the earthquake happened in a country or US state that is on the Ring of Fire
# --- IGNORE ---
ring_of_fire = [
    'Alaska', 'Antarctica', 'Bolivia', 'California', 'Canada', 'Chile', 'Costa Rica', 'Ecuador', 'Fiji',
    'Guatemala', 'Indonesia','Japan', 'Kermadec Islands', 'Mexico',
    'New Zealand', 'Peru', 'Philippines', 'Russia', 'Taiwan', 'Tonga', 'Washington'
]

# Need to remove New Mexico from potentially being flagged as part of the Ring of Fire
df['on_ring_of_fire'] = df['place'].apply(lambda x: any(country in x for country in ring_of_fire) and 'New Mexico' not in x)

# Checking if New Mexico is part of the Ring of Fire
df[['place','on_ring_of_fire']].loc[(df['on_ring_of_fire'] == True) & (df['place'].str.contains('New Mexico', na=False))].count()

place              0
on_ring_of_fire    0
dtype: int64

## Chapter 2, Exercise 5

Calculate the number of earthquakes in the Ring of Fire locations and the number outside of them.

In [50]:
# Calculating the number of earthquake by locations from Ring of Fire and the number outside of them
# Attempted to group by the list of countries/states in the ring_of_fire list
df['ring_of_fire'] = df['place'].apply(
    lambda x: next((country for country in ring_of_fire if country in x), None) if isinstance(x, str) else None
)

# It was at this point I realized that grouping by the list would leave the non-matching values with their full place names
# ring_of_fire_counts = df.groupby('ring_of_fire')['ring_of_fire'].value_counts()

# This approach provides the counts by place and whether it is on the Ring of Fire or not
ring_of_fire_counts = df.groupby('place')['on_ring_of_fire'].value_counts().sort_values(ascending=False)

# This approach provides the counts by whether it is on the Ring of Fire or not using True/False values
ring_of_fire_counts = df['on_ring_of_fire'].value_counts()
ring_of_fire_counts

# # Alternative approach, you can filter the dataframe into two separate dataframes
# ring_of_fire_earthquakes = df[(df['on_ring_of_fire'] == True)]
# outside_ring_of_fire_earthquakes = df[df['on_ring_of_fire'] == False]

# # Have them separated, you can also check the shapes of the two dataframes
# ring_of_fire_earthquakes.shape, outside_ring_of_fire_earthquakes.shape


((4422, 29), (4910, 29))

## Chapter 2, Exercise 6

Find the tsunami count along the Ring of Fire.

In [35]:
# Find the tsunami count along the Ring of Fire
tsunami_counts = df[df['on_ring_of_fire'] == True]['tsunami'].value_counts()

# 0 for no tsunami, 1 for tsunami for places on the Ring of Fire
tsunami_counts

tsunami
0    4379
1      43
Name: count, dtype: int64